File: typescript/src/daily_commit.ts
import { execFileSync } from "child_process";

function run(command, args = []) {
  try {
    console.log(`🚀 Running: ${command} ${args.join(" ")}`);
    execFileSync(command, args, { stdio: "inherit" });
  } catch (error) {
    console.error(`❌ Error running ${command}:`, error.message);
    process.exit(1);
  }
}

function getDate() {
  const now = new Date();
  return now.toISOString().split("T")[0]; // e.g. 2025-11-13
}

function main() {
  const date = getDate();
  const commitMessage = `Daily commit (TypeScript) - ${date}`;

  // Ensure git is clean before committing
  run("git", ["config", "--global", "user.name", "GitHub Actions"]);
  run("git", ["config", "--global", "user.email", "actions@github.com"]);

  // Stage all changes
  run("git", ["add", "."]);

  try {
    // Commit with message (safe way — no shell interpolation)
    run("git", ["commit", "-m", commitMessage]);
  } catch (err) {
    console.log("⚠️ No changes to commit or commit failed:", err.message);
    return;
  }

  // Push changes
  run("git", ["push"]);
  console.log(`✅ Successfully committed and pushed changes for ${date}`);
}

main();

File: dart/bin/daily_commit.dart
import 'dart:io';

String _findRepoRoot() {
  Directory current = Directory.current;
  while (current.path != current.parent.path) {
    final gitDir = Directory('${current.path}${Platform.pathSeparator}.git');
    final readmeFile = File('${current.path}${Platform.pathSeparator}README.md');
    if (gitDir.existsSync() || readmeFile.existsSync()) {
      return current.path;
    }
    current = current.parent;
  }
  // Fallback to current directory
  return Directory.current.path;
}

void main() async {
  try {
    // Find repository root
    final repoRoot = _findRepoRoot();
    Directory.current = Directory(repoRoot);
    
    // Update the timestamp file
    final now = DateTime.now();
    final timestamp = now.toIso8601String();
    final dateStr = '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}';
    
    final timestampPath = '$repoRoot${Platform.pathSeparator}dart${Platform.pathSeparator}timestamp.txt';
    final file = File(timestampPath);
    await file.writeAsString('Last updated: $timestamp\nDate: $dateStr\n');
    
    print('✅ Updated timestamp.txt: $timestamp');
    
    // Configure git (if not already configured)
    await _runGitCommand(['config', 'user.name', 'GitHub Actions']);
    await _runGitCommand(['config', 'user.email', 'actions@github.com']);
    
    // Stage the file
    await _runGitCommand(['add', 'dart/timestamp.txt']);
    print('✅ Staged dart/timestamp.txt');
    
    // Commit the change
    final commitMessage = 'Daily commit (Dart) - $dateStr';
    await _runGitCommand(['commit', '-m', commitMessage]);
    print('✅ Committed changes: $commitMessage');
    
    // Push to repository
    await _runGitCommand(['push']);
    print('✅ Pushed changes to repository');
    
    print('\n🎉 Daily commit completed successfully!');
  } catch (e) {
    print('❌ Error: $e');
    exit(1);
  }
}

Future<void> _runGitCommand(List<String> args) async {
  final result = await Process.run('git', args);
  final stdout = result.stdout.toString();
  final stderr = result.stderr.toString();
  final output = '$stdout$stderr';
  
  // Check for acceptable "nothing to commit" scenarios
  final isNothingToCommit = output.toLowerCase().contains('nothing to commit') ||
      output.toLowerCase().contains('working tree clean') ||
      output.toLowerCase().contains('nothing added to commit');
  
  if (result.exitCode != 0 && !isNothingToCommit) {
    throw Exception('Git command failed: $stderr');
  }
}

